training.qmd

```diff
@@ -1,95 +1,663 @@
+since the matrix-multiplication can be efficiently parallelized on a
+GPU, using standard techniques like blocking and tiling to maximize
+resource utilization. Other alternative ways to accelerate
+matrix-multiplication include using FPGAs, or dedicated hardware like
+TPUs (i.e: systolic arrays).
+
+![AI training is typically compute bound due to the high arithmetic intensity of matrix-multiplication when batch size is large](images/aitrainingroof.png)
+
+## Training Parallelization
+
+Training can be computationally intensive. As outlined above, backpropagation can be expensive both in terms of computation and memory. In terms of computation, backpropagation requires many large matrix multiplications which require considerable arithmetic operation. In terms of memory, backpropagation requires storing the model parameters and intermediate activations in memory, which can be significant if training large models. Handling these difficult system challenges is key to training large models. To mitigate these computational and memory challenges, parallelization is necessary. Broadly, there are two different approaches to parallelizing training: Data Parallel and Model Parallel, discussed below.
+
+### Data Parallel
+
+Data parallelization is a common approach to parallelize machine learning (ML) training across multiple processing units, such as GPUs or distributed computing resources. In data parallelism, the training dataset is divided into batches, and each batch is processed by a separate processing unit. The model parameters are then updated based on the gradients computed from the processing of each batch.
+
+Here's a step-by-step description of data parallel parallelization for ML training:
+
+* Dividing the Dataset
+    The entire training dataset is divided into smaller batches. Each batch contains a subset of the training examples.
+
+* Replicating the Model
+    The ML model (neural network, for example) is replicated across all processing units. Each processing unit has its copy of the model.
+
+* Parallel Computation
+    Each processing unit takes a different batch and computes the forward and backward passes independently. During the forward pass, the model makes predictions on the input data. During the backward pass, gradients are computed for the model parameters using the loss function.
+
+* Gradient Aggregation
+    After processing their respective batches, the gradients from each processing unit are aggregated. Common aggregation methods include summation or averaging of the gradients.
+
+* Parameter Update
+    The aggregated gradients are used to update the model parameters. The update can be performed using optimization algorithms like Stochastic Gradient Descent (SGD) or variants like Adam.
+
+* Synchronization
+    All processing units synchronize their model parameters after the update. This ensures that each processing unit has the latest version of the model.
```

Comment:
Line break between heading and explanation not present when rendered. Might have to add extra line break (like in activation functions section) or `  ` (two spaces) after heading